sharding/networksharding/peerShardMapper.go

```diff
@@ -113,15 +129,86 @@ func (psm *PeerShardMapper) byIDSearchingPidInFallbackCache(pid p2p.PeerID) (sha
 
 	shard, ok := shardObj.(uint32)
 	if !ok {
+		psm.fallbackPidShard.Remove([]byte(pid))
 		return core.UnknownShardId
 	}
 
 	return shard
 }
 
 // UpdatePeerIdPublicKey updates the peer ID - public key pair in the corresponding map
+// It also uses the intermediate pkPeerId cache that will prevent having thousands of peer ID's with
+// the same Elrond PK that will make the node prone to an eclipse attack
 func (psm *PeerShardMapper) UpdatePeerIdPublicKey(pid p2p.PeerID, pk []byte) {
-	psm.peerIdPk.HasOrAdd([]byte(pid), pk)
+	//mutUpdatePeerIdPublicKey is used as to consider this function a critical section
+	psm.mutUpdatePeerIdPublicKey.Lock()
+	defer psm.mutUpdatePeerIdPublicKey.Unlock()
+
+	psm.removePidAssociation(pid)
+
+	objPidsQueue, found := psm.pkPeerId.Get(pk)
+	if !found {
+		psm.peerIdPk.Put([]byte(pid), pk)
+		pq := newPidQueue()
+		pq.push(pid)
+		psm.pkPeerId.Put(pk, pq)
+		return
+	}
+
+	pq, ok := objPidsQueue.(*pidQueue)
+	if !ok {
+		psm.pkPeerId.Remove(pk)
+		return
+	}
+
+	idxPid := pq.indexOf(pid)
+	if idxPid > -1 {
```

Comment:
please create a constant for -1 e.g idxNotFound and check against that  ( != idxNotFound ). 
you can also use maxint and make the type unsigned.

Reply:
done, added const